### ISSUE TITLE
[BUG] worker error: invalid data found in the buffer (possible echo) when using large Json

### ISSUE THREAD

**Issue by maxbaldanza** [spiral/roadrunner#329]
Hi,

We're using roadrunner with docker and PHP but we seem to be getting `worker error: invalid data found in the buffer (possible echo)` when outputting a large json file.
I've tried using the [skeleton docker image](https://github.com/n1215/roadrunner-docker-skeleton) provided and can replicate the issue by changing the `worker.php` to output json like so:

```php
<?php
declare(strict_types=1);

ini_set('display_errors', 'stderr');
include 'vendor/autoload.php';

$relay = new Spiral\Goridge\StreamRelay(STDIN, STDOUT);
$worker = new Spiral\RoadRunner\Worker($relay);
$psr17Factory = new Nyholm\Psr7\Factory\Psr17Factory();
$psr7 = new Spiral\RoadRunner\PSR7Client($worker, $psr17Factory, $psr17Factory, $psr17Factory);

$json = <<<JSON
{
	"@context": {
		"@vocab": "http:\/\/localhost:8000\/docs.jsonld#",
		"hydra": "http:\/\/www.w3.org\/ns\/hydra\/core#",
		"rdf": "http:\/\/www.w3.org\/1999\/02\/22-rdf-syntax-ns#",
		"rdfs": "http:\/\/www.w3.org\/2000\/01\/rdf-schema#",
		"xmls": "http:\/\/www.w3.org\/2001\/XMLSchema#",
		"owl": "http:\/\/www.w3.org\/2002\/07\/owl#",
		"schema": "http:\/\/schema.org\/",
		"domain": {
			"@id": "rdfs:domain",
			"@type": "@id"
		},
		"range": {
			"@id": "rdfs:range",
			"@type": "@id"
		},
		"subClassOf": {
			"@id": "rdfs:subClassOf",
			"@type": "@id"
		},
		"expects": {
			"@id": "hydra:expects",
			"@type": "@id"
		},
		"returns": {
			"@id": "hydra:returns",
			"@type": "@id"
		}
	},
	"@id": "\/docs.jsonld",
	"@type": "hydra:ApiDocumentation",
	"hydra:entrypoint": "\/",
	"hydra:supportedClass": [{
			"@id": "#Import",
			"@type": "hydra:Class",
			"rdfs:label": "Import",
			"hydra:title": "Import",
			"hydra:supportedProperty": [{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#Import\/importType",
						"@type": "rdf:Property",
						"rdfs:label": "importType",
						"domain": "#Import",
						"range": "xmls:string"
					},
					"hydra:title": "importType",
					"hydra:required": true,
					"hydra:readable": true,
					"hydra:writeable": true
				}
			],
			"hydra:supportedOperation": [{
					"@type": [
						"hydra:Operation",
						"schema:FindAction"
					],
					"hydra:method": "GET",
					"hydra:title": "Retrieves Import resource.",
					"rdfs:label": "Retrieves Import resource.",
					"returns": "#Import"
				},
				{
					"@type": [
						"hydra:Operation",
						"schema:DeleteAction"
					],
					"hydra:method": "DELETE",
					"hydra:title": "Deletes the Import resource.",
					"rdfs:label": "Deletes the Import resource.",
					"returns": "owl:Nothing"
				},
				{
					"@type": [
						"hydra:Operation",
						"schema:FindAction"
					],
					"hydra:method": "GET",
					"hydra:title": "Retrieves the collection of ImportLine resources.",
					"rdfs:label": "Retrieves the collection of ImportLine resources.",
					"returns": "#ImportLine"
				}
			]
		},
		{
			"@id": "#OrderUnit",
			"@type": "hydra:Class",
			"rdfs:label": "OrderUnit",
			"hydra:title": "OrderUnit",
			"hydra:supportedProperty": [{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/sku",
						"@type": "rdf:Property",
						"rdfs:label": "sku",
						"domain": "#OrderUnit",
						"range": "xmls:string"
					},
					"hydra:title": "sku",
					"hydra:required": true,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/order",
						"@type": "hydra:Link",
						"rdfs:label": "order",
						"domain": "#OrderUnit",
						"owl:maxCardinality": 1,
						"range": "#Order"
					},
					"hydra:title": "order",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/status",
						"@type": "rdf:Property",
						"rdfs:label": "status",
						"domain": "#OrderUnit",
						"range": "xmls:string"
					},
					"hydra:title": "status",
					"hydra:required": true,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/variant",
						"@type": "rdf:Property",
						"rdfs:label": "variant",
						"domain": "#OrderUnit",
						"range": "xmls:integer"
					},
					"hydra:title": "variant",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/options",
						"@type": "rdf:Property",
						"rdfs:label": "options",
						"domain": "#OrderUnit"
					},
					"hydra:title": "options",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/totalIncTaxBeforeDiscount",
						"@type": "rdf:Property",
						"rdfs:label": "totalIncTaxBeforeDiscount",
						"domain": "#OrderUnit",
						"range": "xmls:integer"
					},
					"hydra:title": "totalIncTaxBeforeDiscount",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/totalExcTaxBeforeDiscount",
						"@type": "rdf:Property",
						"rdfs:label": "totalExcTaxBeforeDiscount",
						"domain": "#OrderUnit",
						"range": "xmls:integer"
					},
					"hydra:title": "totalExcTaxBeforeDiscount",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/totalIncTaxAfterDiscount",
						"@type": "rdf:Property",
						"rdfs:label": "totalIncTaxAfterDiscount",
						"domain": "#OrderUnit",
						"range": "xmls:integer"
					},
					"hydra:title": "totalIncTaxAfterDiscount",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/totalExcTaxAfterDiscount",
						"@type": "rdf:Property",
						"rdfs:label": "totalExcTaxAfterDiscount",
						"domain": "#OrderUnit",
						"range": "xmls:integer"
					},
					"hydra:title": "totalExcTaxAfterDiscount",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/taxBeforeDiscount",
						"@type": "rdf:Property",
						"rdfs:label": "taxBeforeDiscount",
						"domain": "#OrderUnit",
						"range": "xmls:integer"
					},
					"hydra:title": "taxBeforeDiscount",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/taxAfterDiscount",
						"@type": "rdf:Property",
						"rdfs:label": "taxAfterDiscount",
						"domain": "#OrderUnit",
						"range": "xmls:integer"
					},
					"hydra:title": "taxAfterDiscount",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/shippingService",
						"@type": "hydra:Link",
						"rdfs:label": "shippingService",
						"domain": "#OrderUnit",
						"owl:maxCardinality": 1,
						"range": "#ShippingService"
					},
					"hydra:title": "shippingService",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/shippingCost",
						"@type": "rdf:Property",
						"rdfs:label": "shippingCost",
						"domain": "#OrderUnit",
						"range": "xmls:integer"
					},
					"hydra:title": "shippingCost",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/discountTotal",
						"@type": "rdf:Property",
						"rdfs:label": "discountTotal",
						"domain": "#OrderUnit",
						"range": "xmls:integer"
					},
					"hydra:title": "discountTotal",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/discountCodes",
						"@type": "rdf:Property",
						"rdfs:label": "discountCodes",
						"domain": "#OrderUnit"
					},
					"hydra:title": "discountCodes",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				},
				{
					"@type": "hydra:SupportedProperty",
					"hydra:property": {
						"@id": "#OrderUnit\/product",
						"@type": "hydra:Link",
						"rdfs:label": "product",
						"domain": "#OrderUnit",
						"owl:maxCardinality": 1,
						"range": "#Product"
					},
					"hydra:title": "product",
					"hydra:required": false,
					"hydra:readable": true,
					"hydra:writeable": true
				}
			],
			"hydra:supportedOperation": [{
					"@type": [
						"hydra:Operation",
						"schema:FindAction"
					],
					"hydra:method": "GET",
					"hydra:title": "Retrieves OrderUnit resource.",
					"rdfs:label": "Retrieves OrderUnit resource.",
					"returns": "#OrderUnit"
				},
				{
					"@type": [
						"hydra:Operation",
						"schema:ReplaceAction"
					],
					"expects": "#OrderUnit",
					"hydra:method": "PUT",
					"hydra:title": "Replaces the OrderUnit resource.",
					"rdfs:label": "Replaces the OrderUnit resource.",
					"returns": "#OrderUnit"
				}
			]
		}
	]
}
JSON;

while ($req = $psr7->acceptRequest()) {
    try {
        $resp = $psr17Factory->createResponse();
        $resp->getBody()->write($json);
        $psr7->respond($resp);
    } catch (Throwable $e) {
        $psr7->getWorker()->error((string)$e);
    }
}
```

I don't see where I have an echo anywhere and if I make the `json` variable smaller then it doesn't seem to be an issue. 
Strangely enough it doesn't happen for every request so I'm using `ab` to do the requests: `ab -n 1000 -c 10 http://localhost:8080/`:

<img width="1617" alt="Screenshot 2020-05-22 at 12 50 48" src="https://user-images.githubusercontent.com/2786820/82665190-354d3300-9c2b-11ea-823f-0612f35daf4c.png">

Any ideas? I'm not familiar with go so unsure where to look so if you need anything else to help then let me know.

**Comment by 48d90782** [comment]
@maxbaldanza Thanks, we already figured out the problem. New version will be in a moment.

**Comment by 48d90782** [comment]
@maxbaldanza Could you please make `composer update` and re-run this case? Note: Goridge version should be 2.4.3

**Comment by maxbaldanza** [comment]
Thanks for jumping on that so quickly I can now see that I'm using version 2.4.3 of `spiral/goridge` but seem to be getting similar issues:

<img width="904" alt="Screenshot 2020-05-22 at 13 14 22" src="https://user-images.githubusercontent.com/2786820/82666809-8c083c00-9c2e-11ea-949d-b91da026ab7f.png">
<img width="1438" alt="Screenshot 2020-05-22 at 13 15 03" src="https://user-images.githubusercontent.com/2786820/82666988-e3a6a780-9c2e-11ea-8db2-0788e7ada582.png">

**Comment by 48d90782** [comment]
@maxbaldanza You are welcome) Hmm, so, it doesn't relate to the Goridge... I need an extra time to debug the issue.
Have you tried to revert to the RR 1.8.0 ?

**Comment by maxbaldanza** [comment]
Thank you. It does seem that reverting to roadrunner `1.8.0` does solve the issue 👍

**Comment by vvval** [comment]
One question now: what is the version of `goridge` after reverting the `roadrunner`?

**Comment by maxbaldanza** [comment]
Using road runner `1.8.0` with either goridge version `2.4.3` or `2.4.2` solves my issue. 
I can try with another version if it helps you find the issue?

**Comment by 48d90782** [comment]
I'll temporarily remove v1.8.1 tag from RR until we will found and test the issue

**Comment by 48d90782** [comment]
I have pushed the corrected 1.8.1 tag, w/o this bug. But I'll keep this issue opened until I properly investigate the under the hood behavior of this error.

**Comment by 48d90782** [comment]
@maxbaldanza Hey, could you please try to update to RR 1.8.2 version?

**Comment by 48d90782** [comment]
@maxbaldanza fixed in RR 1.8.2. Feel free to reopen the issue if you will see this error again.
